src/qt/bitcoingui.cpp

```diff
@@ -989,6 +989,7 @@ void BitcoinGUI::gotoLoadPSBT(bool from_clipboard)
 
 void BitcoinGUI::updateNetworkState()
 {
+    if (!clientModel) return;
```

Comment:
> Not sure, this could still race, since there is no lock, so clientModel may still be deleted after this statement?

Check it now.
`BitcoinApplication::requestShutdown()` sets the main window (`BitcoinGUI`) `clientModel` to nullptr without clearing the subscribed signals (this is the `window->setClientModel(nullptr)` line), which leaves an open window where the signal handlers are still active but the main window client model is nullptr. So, by unsubscribing the signals prior to the `setClientModel(nullptr)`, we avoid any concurrent event happening in-between the main window nullptr client model and the final clientModel destruction at the end of `BitcoinApplication::requestShutdown()`.

Reply:
> What about events that are already running? Does the destructor ensure they are flushed/discarded, or will it just disconnect?

I haven't looked in detail, but it seems like in principle, if any boost signal events are currently running, and they send asynchronous Q_EMIT qt signals that are run in the GUI event loop thread, there should not be a problem. That is assuming this method is always running the GUI event loop thread, and the code deleting clientModel is also running in the GUI event loop thread.

Reply:
> Ok, but then it means this check in this line could be removed?

I think not if the events are queued, but again I'm not too sure.

Reply:
A bit late but here again. Sorry, life called.

> it means this check in this line could be removed?

Not really. While `requestShutdown()` is being executed, no other task can be processed in the main thread but new tasks can be enqueued to be executed by the main thread after it. So, any event coming from the backend while  `requestShutdown()` is running (prior to the `unsubscribeFromCoreSignals` call) will be executed right after it finishes it.
So.. this means that we also need to guard `RPCConsole::updateNetworkState()` function for a nullptr `clientModel` as well. The same segfault could arise there too..

> I think not if the events are queued, but again I'm not too sure.

Yes. Both events are running on the main thread. The diff is on who enqueues them: the quit action is appended by the main thread (when I press the quit button) and the "network active change" is enqueued by the backend thread, when the RPC changes the value.

Reply:
> if the events all run on a single thread (haven't checked this), this fix should be correct.

Yeah. As far as I have tested, they are running on the main thread. But, this doesn't mean that the code will always act in the same way.. we are using the default qt `connection()` function, which provides `Qt::ConnectionType = Qt::AutoConnection`, which per https://doc.qt.io/qt-6/qt.html#ConnectionType-enum it might use `Qt::DirectConnection`,  which executes the task on the signalling thread.
To solve this, we could force the main thread execution by changing the slot connection side but I think that will leave this investigation for a follow-up. I haven't been able to make it crash again after this fix.